### PR TITLE
Make beatmap carousel select nearest to selection when selection was filtered

### PR DESCRIPTION
**Pull request by LittleEndu**
* Supersedes #14499

This PR changes carousel behaviour when currently selected beatmap gets filtered. Right now the selection is random, this PR would change that to be nearest to previous selection. This PR is basically previous PR but tests fixed.

## Changes
 * Make `BeatmapCarousel` use `PerformSelection` when previous selection was filtered (instead of `SelectNextRandom`)
 * Make `CarouselGroupEagerSelect` perform selection in both direction, selecting nearest to previous selection (instead of only selecting forwards)
 * Remove test case that was testing previous random behaviour 
 * Write new test case for new behaviour

### CODE REVIEW

**Comment by peppy** [comment]
@LittleEndu I've rewritten the main method here to avoid using LINQ, mostly because the logic you had in the LINQ was basically duplicating the non-LINQ logic for the edge case scenarios.

I think this looks good otherwise, but please check my implementation to make sure I haven't (well, the tests haven't) overlooked anything.

**Review by peppy — DISMISSED** [review]

**Comment by LittleEndu** [comment]
Looks good to me.

**Comment by peppy** [comment]
I've fixed the remaining test failure but i'm seeing occasional failures on `TestAddRemove`. I'm not sure if it's related to this change.

JetBrains Rider-EAP 2022-10-25 at 09 59 06

**Comment by LittleEndu** [comment]
I think carousel tests that rely on ordering should specify that beatmaps should be sorted by title. Or sort by title could be made the default. Just saying because I specifically needed to set sort mode to title so that beatmaps are sorted by ID that's in the title.

**Comment by LittleEndu** [comment]
Real failure seems to be that during testing/debugging, carousel root sometimes skips eager selection, and then `TestAddRemove` would succeed. But if it actually performs eager selection, then its last selected index would be bigger than `Items.Count` so the new LINQ free solution would just return null. Changing the line to be `int backwardsIndex = lastSelectedIndex == Items.Count ? Items.Count - 1 : lastSelectedIndex;` would fix and makes sense if deleting last beatmap of the carousel (which the test does)

File: osu.Game/Screens/Select/Carousel/CarouselGroupEagerSelect.cs [hunk 1]
```diff
@@ -108,10 +108,35 @@ private void attemptSelection()
...
+        /// <returns>An unfiltered item nearest to the last selected one or null if all items are filtered</returns>
         protected virtual CarouselItem GetNextToSelect()
         {
-            return Items.Skip(lastSelectedIndex).FirstOrDefault(i => !i.Filtered.Value) ??
-                   Items.Reverse().Skip(Items.Count - lastSelectedIndex).FirstOrDefault(i => !i.Filtered.Value);
+            if (Items.Count == 0)
+                return null;
+
+            int forwardsIndex = lastSelectedIndex;
+            int backwardsIndex = lastSelectedIndex;
```
Comment (LittleEndu) [inline osu.Game/Screens/Select/Carousel/CarouselGroupEagerSelect.cs:121]:
```suggestion
            int backwardsIndex = lastSelectedIndex == Items.Count ? Items.Count - 1 : lastSelectedIndex;
```

**Review by peppy — APPROVED** [review]

**Pull request merged**